plugins/ltac/tacinterp.ml

```diff
@@ -2043,9 +2067,10 @@ let hide_interp {global;ast} =
   let hide_interp env =
     let ist = Genintern.empty_glob_sign ~strict:false env in
     let te = intern_pure_tactic ist ast in
-    let t = eval_tactic te in
-    t
-  in
+    match ast.CAst.v with
+    | TacAtom (TacIntroPattern(false,[ { CAst.v = IntroForthcoming false; loc } ])) ->
+      intro_patterns_deprecated false [CAst.make ?loc @@ IntroForthcoming false] te.CAst.loc
```

Comment:
I don't really like doing this here, how about instead passing the info of being at toplevel down to the interp functions? something like
~~~
modified   plugins/ltac/tacinterp.ml
@@ -1113,14 +1113,14 @@ let rec val_interp ist ?(appl=UnnamedAppl) (tac:glob_tactic_expr) : Val.t Ftacti
   | _ -> value_interp ist >>= fun v -> return (name_vfun appl v)
 
 
-and eval_tactic_ist ist tac : unit Proofview.tactic =
+and eval_tactic_ist ?toplevel ist tac : unit Proofview.tactic =
   let (loc, tac2) = CAst.(tac.loc, tac.v) in
   match tac2 with
   | TacAtom t ->
       let call = LtacAtomCall t in
       let (stack, _) = push_trace(loc,call) ist in
       do_profile stack
-        (catch_error_tac_loc loc stack (interp_atomic ist t))
+        (catch_error_tac_loc loc stack (interp_atomic ?toplevel ist t))
   | TacFun _ | TacLetIn _ | TacMatchGoal _ | TacMatch _ -> interp_tactic ist tac
   | TacId [] -> Proofview.tclLIFT (db_breakpoint (curr_debug ist) [])
   | TacId s ->
@@ -1666,9 +1666,11 @@ and name_atomic ?env tacexpr tac : unit Proofview.tactic =
   Proofview.Trace.name_tactic name tac
 
 (* Interprets a primitive tactic *)
-and interp_atomic ist tac : unit Proofview.tactic =
+and interp_atomic ?toplevel ist tac : unit Proofview.tactic =
   match tac with
   (* Basic tactics *)
+  | TacIntroPattern(false,[ { CAst.v = IntroForthcoming false; loc } ]) when Option.has_some toplevel ->
+    intro_patterns_deprecated false [CAst.make ?loc @@ IntroForthcoming false] toplevel
   | TacIntroPattern (ev,l) ->
       Proofview.Goal.enter begin fun gl ->
         let env = Proofview.Goal.env gl in
@@ -1968,14 +1970,14 @@ let default_ist () =
   let extra = TacStore.set TacStore.empty f_debug (get_debug ()) in
   { lfun = Id.Map.empty; poly = false; extra = extra }
 
-let eval_tactic t =
+let eval_tactic ?toplevel t =
   if get_debug () <> DebugOff then
     Proofview.tclUNIT () >>= fun () -> (* delay for [default_ist] *)
     Proofview.tclLIFT (db_initialize true) <*>
-    eval_tactic_ist (default_ist ()) t
+    eval_tactic_ist ?toplevel (default_ist ()) t
   else
     Proofview.tclUNIT () >>= fun () -> (* delay for [default_ist] *)
-    eval_tactic_ist (default_ist ()) t
+    eval_tactic_ist ?toplevel (default_ist ()) t
 
 let eval_tactic_ist ist t =
   Proofview.tclLIFT (db_initialize false) <*>
@@ -2043,7 +2045,7 @@ let hide_interp {global;ast} =
   let hide_interp env =
     let ist = Genintern.empty_glob_sign ~strict:false env in
     let te = intern_pure_tactic ist ast in
-    let t = eval_tactic te in
+    let t = eval_tactic ?toplevel:te.loc te in
     t
   in
   if global then
~~~

Reply:
It may also be that this warning isn't worth implementing

Reply:
> I don't really like doing this here, how about instead passing the info of being at toplevel down to the interp functions? something like

Indeed, this model while simple, can come back to bite us in the end. Maybe we should have an exception / warning wrapper that can capture the needed closure data, and let the upper layer compute the quickfix too.

This has several advantates:
- matches the lsp performance profile
- is modular and allow people to implement their own QF plugins.

Reply:
I think it is, although not necessarily in this PR